pkg/agent/containerd/watcher.go

```diff
@@ -0,0 +1,359 @@
+	filesMap  map[string]fs.FileInfo
+	workqueue workqueue.TypedRateLimitingInterface[fsnotify.Event]
+}
+
+func CreateWatcher() (*Watcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Watcher{
+		watcher:   watcher,
+		filesMap:  make(map[string]fs.FileInfo),
+		workqueue: workqueue.NewTypedRateLimitingQueue(workqueue.DefaultTypedControllerRateLimiter[fsnotify.Event]()),
+	}, nil
+}
+
+func isFileSupported(path string) bool {
+	for _, ext := range append(tarfile.SupportedExtensions, ".txt") {
+		if strings.HasSuffix(path, ext) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isTarball(path string) bool {
+	for _, ext := range tarfile.SupportedExtensions {
+		if strings.HasSuffix(path, ext) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (w *Watcher) HandleWatch(path string) error {
+	if err := w.watcher.Add(path); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to watch from %s directory: %v", path, err))
+	}
+
+	return nil
+}
+
+// Populate the state of the files in the directory
+// for the watcher to have infos about the file changing
+// this function need to break
+func (w *Watcher) Populate(path string) error {
+	var errs []error
+
+	fileInfos, err := os.ReadDir(path)
+	if err != nil {
+		logrus.Errorf("Unable to read files in %s: %v", path, err)
+		return err
+	}
+
+	for _, dirEntry := range fileInfos {
+		if dirEntry.IsDir() {
+			continue
+		}
+
+		// get the file info to add to the state map
+		fileInfo, err := dirEntry.Info()
+		if err != nil {
+			logrus.Errorf("Failed while getting the info from file: %v", err)
+			errs = append(errs, err)
+			continue
+		}
+
+		if isFileSupported(dirEntry.Name()) {
+			// insert the file into the state map that will have the state from the file
+			w.filesMap[filepath.Join(path, dirEntry.Name())] = fileInfo
+		}
+	}
+
+	return merr.NewErrors(errs...)
+}
+
+func (w *Watcher) ClearMap() {
+	w.filesMap = make(map[string]fs.FileInfo)
+}
+
+func (w *Watcher) runWorkerForImages(ctx context.Context, cfg *config.Node) {
+	// create the connections to not create every time when processing a event
+	client, err := Client(cfg.Containerd.Address)
+	if err != nil {
+		logrus.Errorf("Failed to create containerd client: %v", err)
+		w.watcher.Close()
+		return
+	}
+
+	defer client.Close()
+
+	criConn, err := cri.Connection(ctx, cfg.Containerd.Address)
+	if err != nil {
+		logrus.Errorf("Failed to create CRI connection: %v", err)
+		w.watcher.Close()
+		return
+	}
+
+	defer criConn.Close()
+
+	imageClient := runtimeapi.NewImageServiceClient(criConn)
+
+	for w.processNextEventForImages(ctx, cfg, client, imageClient) {
+	}
+}
+
+func (w *Watcher) processNextEventForImages(ctx context.Context, cfg *config.Node, client *containerd.Client, imageClient runtimeapi.ImageServiceClient) bool {
+	event, shutdown := w.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	if isTarball(event.Name) {
+		time.Sleep(1 * time.Second)
+	}
```

Comment:
Don't sleep in the queue processor, this is not the correct way to introduce a delay to wait for additional events to come in.

Change the workqueue from `TypedRateLimitingQueue` to `DelayingQueue`, and use `workqueue.AddAfter(filePath, duration)` to enqueue changes to events. You'll still need to maintain a map of path to fileinfo, but use of a delaying queue will ensure that subsequent events for a single file delay processing of that file until after no events are received for the specified duration. Just sleeping in the queue handler will NOT do that.